src/Effectful/HPQTypes.hs

```diff
@@ -71,72 +79,128 @@ runEffectDB ::
-    runWithNewConnection ::
-      Eff (State (PQ.DBState (Eff es)) : es) b ->
-      Eff (State (PQ.DBState (Eff es)) : es) b
-    runWithNewConnection action = do
-      dbState :: PQ.DBState (Eff es) <- get
-      result <- PQ.withConnection (PQ.unConnectionSource connectionSource) $ \newConn -> do
-        -- TODO: We do not pass the current connection source to the new DB
-        -- state, which differs from the original code.
-        put $ mkDBState newConn (PQ.dbTransactionSettings dbState)
-        action
-      put dbState
-      pure result
-    mkDBState :: PQ.Connection -> PQ.TransactionSettings -> PQ.DBState (Eff es)
-    mkDBState conn ts =
-      PQ.DBState
-        { PQ.dbConnection = conn
-        , PQ.dbConnectionSource = PQ.unConnectionSource connectionSource
-        , PQ.dbTransactionSettings = ts
-        , PQ.dbLastQuery = PQ.SomeSQL (mempty :: PQ.SQL)
-        , PQ.dbRecordLastQuery = True
-        , PQ.dbQueryResult = Nothing
-        }
+        let dbState0 = mkDBState (PQ.unConnectionSource connectionSource) conn transactionSettings
+        evalState dbState0 $ handleAutoTransaction transactionSettings withTransaction' eff
+    withTransaction' ::
+      PQ.TransactionSettings ->
+      Eff (DBState es : es) a ->
+      Eff (DBState es : es) a
+    withTransaction' ts eff = unDBEff . PQ.withTransaction' ts $ DBEff eff
+
+mkDBState ::
+  PQ.ConnectionSourceM m ->
+  PQ.Connection ->
+  PQ.TransactionSettings ->
+  PQ.DBState m
+mkDBState connectionSource conn ts =
+  PQ.DBState
+    { PQ.dbConnection = conn
+    , PQ.dbConnectionSource = connectionSource
+    , PQ.dbTransactionSettings = ts
+    , PQ.dbLastQuery = PQ.SomeSQL (mempty :: PQ.SQL)
+    , PQ.dbRecordLastQuery = True
+    , PQ.dbQueryResult = Nothing
+    }
+
+handleAutoTransaction ::
+  PQ.TransactionSettings ->
+  (PQ.TransactionSettings -> m a -> m a) ->
+  m a ->
+  m a
+handleAutoTransaction transactionSettings withTransaction action =
+  -- REVIEW: Why don't we have to set `tsAutoTransaction` to `False` in the
+  -- context of the `action`?
+  if PQ.tsAutoTransaction transactionSettings
+    then withTransaction (transactionSettings {PQ.tsAutoTransaction = False}) action
+    else action
+
+---------------------------------------------------
+-- Internal effect stack
+---------------------------------------------------
+
+-- | Newtype wrapper over the internal DB effect stack
+newtype DBEff es a = DBEff
+  { unDBEff :: Eff (DBState es : es) a
+  }
+  deriving newtype (Functor, Applicative, Monad, MonadThrow, MonadCatch, MonadMask)
+
+-- | Internal state effect used to reinterpret the `DB` effect
+type DBState es = State (PQ.DBState (DBEff es))
+
+-- Convenience @MonadBase IO@ instance
+instance (IOE :> es) => MonadBase IO (DBEff es) where
+  liftBase b = DBEff $ liftBase b
+
+-- Convenience instance to avoid writing @DBEff get@ etc.
+-- REVIEW: This comes with the mtl dependency, so maybe it isn't worth it?  The
```

Comment:
Yeah, I don't think it's worth it, I'd suggest removing it.

Reply:
Alright, removed.  I added some helper functions (`get`, `put`, ...) instead.